Include cost estimates in IO EXPLAIN plan

JamesRTaylor opened the pull request:
Implementation of https://github.com/prestosql/presto/issues/677

kokosing commented on presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java:
```diff
@@ -86,6 +91,7 @@ private String print(PlanNode plan)
     {
         private final Set<TableColumnInfo> inputTableColumnInfos;
         private final Optional<CatalogSchemaTableName> outputTable;
+        private final Long estBytesScanned;
```
use Optional<Long> to catch the nullness of this field

kokosing commented on presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java:
```diff
@@ -94,6 +100,22 @@ public IoPlan(
         {
             this.inputTableColumnInfos = ImmutableSet.copyOf(requireNonNull(inputTableColumnInfos, "inputTableColumnInfos is null"));
             this.outputTable = requireNonNull(outputTable, "outputTable is null");
+            long totalEstBytesScanned = 0;
```
what is `Est`? `Estimated`? Please don't use abbreviations.

kokosing commented on presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java:
```diff
@@ -94,6 +100,22 @@ public IoPlan(
+            boolean hasTotalEstBytesScanned = false;
+            for (TableColumnInfo info : inputTableColumnInfos) {
+                Long estBytesScanned = info.getEstBytesScanned();
+                if (estBytesScanned != null) {
```
what if some `TableColumnInfo` have this field and some don't? Does it make sense to calculate `totalEstBytesScanned `, is it accurate enough?

JamesRTaylor replied:
For our use case, the purpose of the estimates output are to get a rough idea of the relative expense of the query The estimate are purely based on statistics generated by the connectors, hence their accuracy cannot be guaranteed. It's really more of a best effort.

kokosing commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTableStatistics.java:
```diff
@@ -756,6 +773,39 @@ public void testAnalyzeForTablePartitionedByBigint()
+                row("p_comment", 1197.0, 5.0, 0.0, null, null, null),
+                row(null, null, null, null, 15.0, null, null));
+
+        String joinQuery = "EXPLAIN (TYPE IO) SELECT n_name, p_name FROM " + tableName1InDatabase + " JOIN " + tableName2InDatabase + " ON (n_nationkey = p_nationkey) WHERE n_comment='foo'";
```
use `String.format`

findepi replied:
note: when using `String.format` we static import the method

kokosing commented on presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java:
```diff
@@ -470,11 +501,33 @@ public Void visitTableScan(TableScanNode node, IoPlanBuilder context)
         {
             TableMetadata tableMetadata = metadata.getTableMetadata(session, node.getTable());
             TupleDomain<ColumnHandle> predicate = metadata.getTableProperties(session, node.getTable()).getPredicate();
+            QualifiedObjectName tableName = new QualifiedObjectName(tableMetadata.getCatalogName().getCatalogName(),
```
move each parameter into separate line like:
```
new SomeClass(
  a,
  b,
  c);
```

kokosing commented on presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java:
```diff
@@ -470,11 +501,33 @@ public Void visitTableScan(TableScanNode node, IoPlanBuilder context)
+                Set<ColumnHandle> predicateColumns = predicate.getDomains().get().keySet();
+                Collection<ColumnHandle> projectedColumns = node.getAssignments().values();
+                for (Map.Entry<ColumnHandle, ColumnStatistics> entry : tableStatistics.getColumnStatistics().entrySet()) {
+                    if (!entry.getValue().getDataSize().isUnknown()
```
IMO if any stats for column is unknown we should mark `estBytesScanned` as unknown as well.

JamesRTaylor replied:
See comment above.

kokosing commented on presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java:
```diff
@@ -470,11 +501,33 @@ public Void visitTableScan(TableScanNode node, IoPlanBuilder context)
+                Constraint constraint = new Constraint(predicate);
+                TableStatistics tableStatistics = metadata.getTableStatistics(session, table, constraint);
+                long bytesScanned = 0;
+                // Only include stats for columns that are projected or used in a filter
```
not needed comment

kokosing commented on presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java:
```diff
@@ -470,11 +501,33 @@ public Void visitTableScan(TableScanNode node, IoPlanBuilder context)
+                    tableMetadata.getTable().getTableName());
+            Optional<TableHandle> optTableHandle = metadata.getTableHandle(session, tableName);
+            Long estBytesScanned = null;
+            if (optTableHandle.isPresent()) {
```
can you extract this whole `if` block as a method?

ebyhr commented:
While I already sent PR https://github.com/prestosql/presto/pull/709 (please ignore mine, I'll close it), why only estimated byte scanned? I think other information (e.g. outputRowCount) is also useful.

And if you add estBytesScanned to TableColumnInfo directly, the output format will be like below.
```
{
  "inputTableColumnInfos": [
    {
      "table": {..
      "columns": [ ..
      "estimatedBytesScanned": 1230000
    }
  ],
```

While it's easy to parse for now, when we add fields later about estimated costs, it looks little redundant to me.
```
{
  "inputTableColumnInfos": [
    {
      "table": {
      "columns": [ ..
      "estimatedBytesScanned": 1230000
      "estimatedoutputRowCount": 999
    }
  ],
```

I hope you update presto-docs/src/main/sphinx/sql/explain.rst so that other developers can use this enhancement easily.

findepi commented:
> While I already sent PR #709 ...

@ebyhr apologies it hasn't been reviewed yet!

@ebyhr @JamesRTaylor can you look at each other's PR and determine how much overlap is there?

ebyhr commented:
@findepi There's no need to apologize. Actually, I've completely forgot my PR until today and I hope @JamesRTaylor (the issue ticket was created by him) implements this. 

The main difference between my PR and this commit is I added a separated class and 5 metrics including estimated byte scanned. If there's no special reason to limit to the bytes, adding other 4 metrics looks better to me.

JamesRTaylor commented:
Thanks so much for your comments, @kokosing, @ebyhr, and @findepi. I've pushed an update to address the feedback. Please review when you get a chance.

JamesRTaylor commented:
Ping @kokosing, @dain, @martint - please review when you have a few spare cycles. Thanks!

kokosing commented:
Can you please rebase and get rid of merge commits (we do not use them).

JamesRTaylor commented:
@kokosing - I rebased and squashed the commits. I looked at the test failure, but it doesn't appear related.

findepi commented:
@JamesRTaylor from Travis:
```
[ERROR]   TestTpchDistributedQueries.testIOExplain:70 expected [IoPlan{inputTableColumnInfos=[TableColumnInfo{table=tpch.sf0.01.orders, columnConstraints=[ColumnConstraint{columnName=orderstatus, typeSignature=varchar(1), domain=FormattedDomain{nullsAllowed=false, ranges=[FormattedRange{low=FormattedMarker{value=Optional[F], bound=EXACTLY}, high=FormattedMarker{value=Optional[F], bound=EXACTLY}}, FormattedRange{low=FormattedMarker{value=Optional[O], bound=EXACTLY}, high=FormattedMarker{value=Optional[O], bound=EXACTLY}}, FormattedRange{low=FormattedMarker{value=Optional[P], bound=EXACTLY}, high=FormattedMarker{value=Optional[P], bound=EXACTLY}}]}}], estimate=EstimatedStatsAndCost{outputRowCount=15000.0, outputSizeInBytes=1597294.0, cpuCost=1597294.0, maxMemory=0.0, networkCost=0.0, bytesScanned=0.0}}], outputTable=Optional.empty, estimate=EstimatedStatsAndCost{outputRowCount=15000.0, outputSizeInBytes=1597294.0, cpuCost=1597294.0, maxMemory=0.0, networkCost=0.0, bytesScanned=0.0}}] but found [IoPlan{inputTableColumnInfos=[TableColumnInfo{table=tpch.sf0.01.orders, columnConstraints=[ColumnConstraint{columnName=orderstatus, typeSignature=varchar(1), domain=FormattedDomain{nullsAllowed=false, ranges=[FormattedRange{low=FormattedMarker{value=Optional[F], bound=EXACTLY}, high=FormattedMarker{value=Optional[F], bound=EXACTLY}}, FormattedRange{low=FormattedMarker{value=Optional[O], bound=EXACTLY}, high=FormattedMarker{value=Optional[O], bound=EXACTLY}}, FormattedRange{low=FormattedMarker{value=Optional[P], bound=EXACTLY}, high=FormattedMarker{value=Optional[P], bound=EXACTLY}}]}}], estimate=EstimatedStatsAndCost{outputRowCount=15000.0, outputSizeInBytes=1597294.0, cpuCost=1597294.0, maxMemory=0.0, networkCost=0.0, bytesScanned=742294.0}}], outputTable=Optional.empty, estimate=EstimatedStatsAndCost{outputRowCount=15000.0, outputSizeInBytes=1597294.0, cpuCost=1597294.0, maxMemory=0.0, networkCost=1597294.0, bytesScanned=742294.0}}]
```

JamesRTaylor commented:
Oops, missed that - sorry about that. Was seeing other errors in the console that I thought was leading to the test failure. I'll get that fixed.

JamesRTaylor commented:
All fixed now, @findepi. Sorry again for the noise.

kokosing commented on presto-docs/src/main/sphinx/sql/explain.rst:
```diff
@@ -139,14 +139,30 @@ IO:
                }
```
commit message is a bit too long

kokosing commented on presto-docs/src/main/sphinx/sql/explain.rst:
```diff
@@ -139,14 +139,30 @@ IO:
+         } ],
+         "estimate" : {
+           "outputRowCount" : 25.0,
+           "outputSizeInBytes" : 2734.0,
```
Can you please give an example with higher values? It would be nice to use units here (instead 1000000 use 1M, or 1MB in case of bytes).

kokosing commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -233,52 +239,78 @@ public void testIOExplain()
+        EstimatedStatsAndCost estimate16 = new EstimatedStatsAndCost(1.0, 16.0, 16.0, 0.0, 0.0, 6.0);
+        EstimatedStatsAndCost estimate17 = new EstimatedStatsAndCost(1.0, 17.0, 17.0, 0.0, 0.0, 3.0);
+        EstimatedStatsAndCost estimate25 = new EstimatedStatsAndCost(1.0, 25.0, 25.0, 0.0, 0.0, 3.0);
+        Map<Object, TypeAndEstimate> data = new LinkedHashMap<>();
```
why linked hash map?

JamesRTaylor replied:
It's a linked hash map because the assert below prints the index of the entry upon failure to make the entry easier to find.

findepi replied:
We usually use `ImmutableMap.Builder` also when we want to preserve insertion order.
ImmutableMap guarantees order.

BTW here `List<Entry<..>` would do. Or restructuring the test so that `data` is not needed at all.

kokosing commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -233,52 +239,78 @@ public void testIOExplain()
+        EstimatedStatsAndCost estimate13 = new EstimatedStatsAndCost(1.0, 13.0, 13.0, 0.0, 0.0, 3.0);
+        EstimatedStatsAndCost estimate16 = new EstimatedStatsAndCost(1.0, 16.0, 16.0, 0.0, 0.0, 6.0);
+        EstimatedStatsAndCost estimate17 = new EstimatedStatsAndCost(1.0, 17.0, 17.0, 0.0, 0.0, 3.0);
+        EstimatedStatsAndCost estimate25 = new EstimatedStatsAndCost(1.0, 25.0, 25.0, 0.0, 0.0, 3.0);
```
please inline these variables

JamesRTaylor replied:
These values are used multiple times, so I made them variables.

kokosing commented on presto-main/src/main/java/io/prestosql/sql/analyzer/QueryExplainer.java:
```diff
@@ -122,7 +122,8 @@ public String getPlan(Session session, Statement statement, Type planType, List<
                 SubPlan subPlan = getDistributedPlan(session, statement, parameters, warningCollector);
                 return PlanPrinter.textDistributedPlan(subPlan, metadata, session, false);
             case IO:
-                return IoPlanPrinter.textIoPlan(getLogicalPlan(session, statement, parameters, warningCollector).getRoot(), metadata, session);
```
undo formatting change (inline `ioPlan`)

kokosing commented on presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java:
```diff
@@ -514,6 +671,41 @@ else if (writerTarget instanceof CreateReference || writerTarget instanceof Inse
+            if (optTableHandle.isPresent()) {
+                TableHandle table = optTableHandle.get();
+                Constraint constraint = new Constraint(predicate);
+                TableStatistics tableStatistics = metadata.getTableStatistics(session, table, constraint);
```
Getting table statistics could be potentially expensive. It would be nice to leverage only information from `StatsAndCosts`. Could first expose estimates without `bytesScanned` statistic.

Also why `stats.getOutputSizeInBytes()` is not enough? What is the difference between `stats.getOutputSizeInBytes()` and `bytesScanned`? IMO they are the same thing.

kokosing commented on presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java:
```diff
@@ -514,6 +671,41 @@ else if (writerTarget instanceof CreateReference || writerTarget instanceof Inse
+                    tableMetadata.getTable().getSchemaName(),
+                    tableMetadata.getTable().getTableName());
+            Optional<TableHandle> optTableHandle = metadata.getTableHandle(session, tableName);
+            double bytesScanned = 0;
```
* can you please extract the below if block as method (getBytesScannedStatistics). 
* Also in case where there is no table handle or no stats, please use `Double.NaN`. Using `0` could be misleading.

kokosing commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTableStatistics.java:
```diff
@@ -756,6 +775,42 @@ public void testAnalyzeForTablePartitionedByBigint()
+        String tableName2InDatabase = mutableTablesState().get(NATION_PARTITIONED_BY_BIGINT_REGIONKEY.getName()).getNameInDatabase();
+
+        String showStatsWholeTable1 = "SHOW STATS FOR " + tableName1InDatabase;
+        String showStatsWholeTable2 = "SHOW STATS FOR " + tableName2InDatabase;
```
inline `showStatsWholeTable*` variables

kokosing commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTableStatistics.java:
```diff
@@ -14,6 +14,7 @@
 package io.prestosql.tests.hive;
 
 import com.google.common.collect.ImmutableList;
+import io.prestosql.sql.planner.planprinter.IoPlanPrinter;
```
by general we tend to not to use presto main code in product tests. However this usage seems reasonable.

kokosing commented on presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java:
```diff
@@ -514,6 +671,41 @@ else if (writerTarget instanceof CreateReference || writerTarget instanceof Inse
+                Set<ColumnHandle> predicateColumns = predicate.getDomains().get().keySet();
+                Collection<ColumnHandle> projectedColumns = node.getAssignments().values();
+                for (Map.Entry<ColumnHandle, ColumnStatistics> entry : tableStatistics.getColumnStatistics().entrySet()) {
+                    if (!entry.getValue().getDataSize().isUnknown()
```
if one value is unknown then result is also unknown.

JamesRTaylor commented:
@kokosing - thanks for the prior review. I've updated the PR based on your feedback:
removed estimatedBytesScanned as it wouldn't be accurate anyway.
squashed the commits and shortened the commit message

kokosing commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -233,52 +239,78 @@ public void testIOExplain()
-        data.put("2019-01-01 23:22:21.123", TimestampType.TIMESTAMP);
-        for (Map.Entry<Object, Type> entry : data.entrySet()) {
-            @Language("SQL") String query = format(
+        EstimatedStatsAndCost estimate10 = new EstimatedStatsAndCost(1.0, 10.0, 10.0, 0.0, 0.0);
```
please inline these variables

kokosing commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -233,52 +239,78 @@ public void testIOExplain()
+                    Optional.empty(),
+                    estimate);
+            IoPlan actualPlan = jsonCodec(IoPlan.class).fromJson((String) getOnlyElement(result.getOnlyColumnAsSet()));
+            assertEquals(actualPlan, expectedPlan, format("%d) Type %s ", index, type));
```
please inline `actualPlan` and `expectedPlan` as it used to be

kokosing commented on presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java:
```diff
@@ -71,30 +78,33 @@ private IoPlanPrinter(Metadata metadata, Session session)
         {
             this.inputTableColumnInfos = ImmutableSet.copyOf(requireNonNull(inputTableColumnInfos, "inputTableColumnInfos is null"));
-            this.outputTable = requireNonNull(outputTable, "outputTable is null");
+            this.outputTable = outputTable;
```
requireNonNull here and below

kokosing commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTableStatistics.java:
```diff
@@ -687,9 +687,9 @@ public void testAnalyzeForTablePartitionedByBigint()
-        String showStatsPartitionTwo = "SHOW STATS FOR (SELECT * FROM " + tableNameInDatabase + " WHERE p_regionkey = 2)";
+        String showStatsWholeTable = format("SHOW STATS FOR %s", tableNameInDatabase);
+        String showStatsPartitionOne = format("SHOW STATS FOR (SELECT * FROM %s WHERE p_regionkey = 1)", tableNameInDatabase);
+        String showStatsPartitionTwo = format("SHOW STATS FOR (SELECT * FROM %s WHERE p_regionkey = 2)", tableNameInDatabase);
```
please extract these changes as separate commit, it is not much related to the main change you are introducing. This comment applies also to other such changes in this file.

kokosing commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTableStatistics.java:
```diff
@@ -734,7 +735,7 @@ public void testAnalyzeForTablePartitionedByBigint()
         // analyze for all partitions
 
-        assertThat(query("ANALYZE " + tableNameInDatabase)).containsExactly(row(15));
+        assertThat(query(format("ANALYZE %s", tableNameInDatabase))).containsExactly(row(15));
```
undo. Our convention is to not to use `format` where it can be replaced with simple concatenation like `"message: " + variable`

kokosing commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTableStatistics.java:
```diff
@@ -758,6 +759,31 @@ public void testAnalyzeForTablePartitionedByBigint()
 
+    @Test
+    @Requires({NationPartitionedByBigintTable.class, UnpartitionedNationTable.class})
+    public void testEstBytesScannedAcrossJoinedTables()
```
s/Est/Esimated

kokosing commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTableStatistics.java:
```diff
@@ -758,6 +759,31 @@ public void testAnalyzeForTablePartitionedByBigint()
+    @Requires({NationPartitionedByBigintTable.class, UnpartitionedNationTable.class})
+    public void testEstBytesScannedAcrossJoinedTables()
+    {
+        String tableName1InDatabase = mutableTablesState().get(NATION.getName()).getNameInDatabase();
```
can you have two tests for this, one per each table. It looks like these tables are not used together. How do they relate?

kokosing commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTableStatistics.java:
```diff
@@ -758,6 +759,31 @@ public void testAnalyzeForTablePartitionedByBigint()
+                row("n_name", 177.0, 25.0, 0.0, null, null, null),
+                row("n_regionkey", null, 5.0, 0.0, null, "0", "4"),
+                row("n_comment", 1857.0, 25.0, 0.0, null, null, null),
+                row(null, null, null, null, 25.0, null, null));
```
what changed here? How this assertion would look like before your change in the product?

kokosing commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTableStatistics.java:
```diff
@@ -707,9 +707,10 @@ public void testAnalyzeForTablePartitionedByBigint()
                 row("p_comment", null, null, null, null, null, null),
                 row(null, null, null, null, null, null, null));
 
+        String selectStarAllPartitions = format("EXPLAIN (TYPE IO) SELECT * FROM %s", tableNameInDatabase);
```
this variable not used

JamesRTaylor commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -210,14 +213,17 @@ public void testIOExplain()
-                                                                        new FormattedMarker(Optional.of("false"), EXACTLY)))))))),
-                        Optional.of(new CatalogSchemaTableName(catalog, "tpch", "test_orders"))));
+                                                                        new FormattedMarker(Optional.of("false"), EXACTLY)))))),
+                                estimate)),
```
There's definitely a lot of nesting here. The estimate variable is an argument of the new TableColumnInfo() call and is indented at the same level as the other arguments. Where should it be if not there?

JamesRTaylor replied:
Or perhaps it's best to move the new TableColumnInfo() call to it's own line and intent as you've indicated (though there'll be more whitespace diffs then)?

JamesRTaylor commented:
@kokosing - thanks for your previous reviews. I've pushed another commit based on your feedback.

kokosing commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -210,15 +213,18 @@ public void testIOExplain()
 
-        result = computeActual("EXPLAIN (TYPE IO, FORMAT JSON) INSERT INTO test_orders SELECT custkey, orderkey + 10 FROM test_orders WHERE custkey <= 10");
+        estimate = new EstimatedStatsAndCost(55.0, 990.0, 990.0, 0.0, 0.0);
+        result = computeActual("EXPLAIN (TYPE IO, FORMAT JSON) INSERT INTO test_orders SELECT custkey, orderkey + 10 FROM test_orders where custkey <= 10");
```
where -> WHERE

kokosing commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -233,52 +239,68 @@ public void testIOExplain()
+                                                                        new FormattedMarker(Optional.of("199"), EXACTLY)))))),
+                                estimate)),
+                        Optional.of(new CatalogSchemaTableName(catalog, "tpch", "test_orders")),
+                        estimate));
```
````
methodCall(
     argument1,
     argument2,
     argument3);
```

kokosing commented on presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java:
```diff
@@ -471,12 +606,14 @@ public Void visitTableScan(TableScanNode node, IoPlanBuilder context)
                             tableMetadata.getTable().getTableName()),
-                    parseConstraints(node.getTable(), predicate)));
+                    parseConstraints(node.getTable(), predicate),
+                    estimatedStatsAndCost));
```
don't wrap the line here

kokosing commented on presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java:
```diff
@@ -514,6 +651,26 @@ else if (writerTarget instanceof CreateReference || writerTarget instanceof Inse
+                    tableMetadata.getCatalogName().getCatalogName(),
+                    tableMetadata.getTable().getSchemaName(),
+                    tableMetadata.getTable().getTableName());
+            Optional<TableHandle> optTableHandle = metadata.getTableHandle(session, tableName);
```
s/optTableHandle/tableHandle

Also `optTableHandle` looks not used, and this could result in not needed metastore call, please remove.

kokosing commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -183,102 +185,118 @@ public void testIOExplain()
-                                                                        new FormattedMarker(Optional.of("false"), EXACTLY),
-                                                                        new FormattedMarker(Optional.of("false"), EXACTLY)))))))),
-                        Optional.of(new CatalogSchemaTableName(catalog, "tpch", "test_orders"))));
+                        ImmutableSet.of(
```
now when you change the formatting of existing code that is not strictly related to your change, please extract that as separate commit. Before your change.

sorry about such comments

kokosing commented on presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java:
```diff
@@ -471,12 +471,13 @@ public Void visitTableScan(TableScanNode node, IoPlanBuilder context)
+                                tableMetadata.getCatalogName().getCatalogName(),
+                                tableMetadata.getTable().getSchemaName(),
+                                tableMetadata.getTable().getTableName()),
+                        parseConstraints(node.getTable(), predicate)));
```
❤️ Thank you!

kokosing commented:
It looks there some issue with travis, some external service was unavailable. Restarted.

kokosing commented:
Can you please raise `Xmx` in `MAVEN_OPTS` in `.travis.yml` to `1G`.

FYI: @electrum @findepi

findepi commented on presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -183,102 +185,118 @@ public void testIOExplain()
+                                    estimate)),
+                            Optional.empty(),
+                            estimate),
+                    format("%d) Type %s ", index, type));
```
missing `(` ?

findepi commented:
> Can you please raise `Xmx` in `MAVEN_OPTS` in `.travis.yml` to `1G`.

So far we've been seeing OOM on full maven build (with verification), and rather rarely.
This PR OOMs during compilation several times (https://travis-ci.com/prestosql/presto/builds/124792147). I don't see any reason why it would be so, but that's strange.

@JamesRTaylor could you please trigger a _new_ build (rather than retriggering the failed one) so that we can see how reproducible it is. You can also trigger a build with changed `MAVEN_OPTS`.

eg
- push an empty commit
- push a commit changing `MAVEN_OPTS`
- go to Travis and make sure the empty commit's build is run (Travis auto-cancels builds but you can still manually force a build to run)

@kokosing We can probably increase -Xmx for Maven, but need to think whether we don't go over limit somewhere else, when maven is running concurrently. 
Like tests execution in a forked JVM (but tests are limited at 2g). Anything else?

kokosing commented:
>  I don't see any reason why it would be so, but that's strange.

It is strange to me as well. Let me retrigger the build with flushed cache.

findepi commented:
> It is strange to me as well. Let me retrigger the build with flushed cache.

i see it green, so ... @JamesRTaylor @kokosing let's do nothing and wait until problem resurfaces.

kokosing commented:
Merged, thanks!
